### PR TITLE
Dice loss modified to eliminate the batch size affects (issue #910)

### PR DESCRIPTION
**Pull request by gbazad93**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
The dice loss function was modified to eliminate the batch size effect. When the batch contains a sample with a tiny target, its contribution to the loss function is comparatively less than a sample with a big object, while each instance should have the same weight on the loss function. Thus, the current pull request aims to address this issue. 

## Linked issues
[##910](https://github.com/ivadomed/ivadomed/issues/910)

### CODE REVIEW

**Comment by hermancollin** [comment]
I think this is an important feature. I am currently re-running a training I did a few weeks ago using this PR and I am very curious to see if it will affect the model's performance. I predict it will, because my dataset contains a few near-empty samples.

**Comment by gbazad93** [comment]
> 
> 
> I think this is an important feature. I am currently re-running a training I did a few weeks ago using this PR and I am very curious to see if it will affect the model's performance. I suspect it will, because my dataset contains a few near-empty samples.
> 
> update: unfortunately, my model does not perform as well as before. FWI I used MultiClassDiceLoss (wrapper over the dice loss class)

Hi, @hermancollin thanks for your experiment. Could you please specify your batch size? And by performance, do you mean results on the validation or test set? In the case of data set with relatively small or empty target samples, I wonder if the modification might be responsible for not complete convergence. Training and validation loss diagrams might be a helpful metric to observe this behavior.

**Comment by charleygros** [comment]
Implementation looks good to me, many thanks @gbazad93, great contribution!

I wonder if we should keep this (eg average per sample or per batch) as a parameter of the Loss function, ie let the user choose how the loss is computed. I feel it's really task-specific. We could have a default behviour of course.
What do you think @gbazad93 / @andreanne-lemay / @mariehbourget / @uzaymacar ? Cheers

**Comment by gbazad93** [comment]
@charleygros, thanks for the update. In a medical case, the ultimate objective is to produce the segmentation mask for an individual case. Thus, sample-wise averaging might provide more realistic results than batch-wise. On the other hand, averaging the batch can provide a smooth convergence and is useful for training. With further scrutinized the [Pytorch loss implementations](https://github.com/pytorch/pytorch/blob/22f78144c70d9d215bc390805aa8d0657e548b85/torch/nn/functional.py) I found that Pytorch uses average over batch. Have a look at the following experiments:
image

So, as @charleygros suggested it can be optional for specific dataset and necessary for the test set.

**Comment by hermancollin** [comment]
> Hi, @hermancollin thanks for your experiment. Could you please specify your batch size? And by performance, do you mean results on the validation or test set? In the case of data set with relatively small or empty target samples, I wonder if the modification might be responsible for not complete convergence. Training and validation loss diagrams might be a helpful metric to observe this behavior.

I used a batch size of 6. The results were worst on both validation and testing sets. To be fair, I suspect the hyperparameters were optimized for batch-wise averaging. Obtaining the same results as before with this PR should be possible but might require more optimization.

I am also in favor of making this feature optional. Batch-wise averaging seems to work well currently with default architectures.

**Comment by gbazad93** [comment]
@hermancollin Thanks for the update. My initial issue was for the performance on the validation set. For the training purpose I think the evidence shows that the  batch-wise normalization can provide a better generalization performance. Thanks again for your experiment.

**Comment by hermancollin** [comment]
@gbazad93 Ah I understand the issue now. In that case, I agree that computing sample-wise averaged Dice on the test set makes sense.

**Comment by charleygros** [comment]
> So, as @charleygros suggested it can be optional for specific dataset and necessary for the test set.

Agree with this suggestion, let's do it :-)

Note that we have something similar already implemented at test time: the [evaluation pipeline on the 3D reconstructed images](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/evaluation.py#L114) --> which computes the evaluation metrics on the full 3D MRI image, instead of "per batch" or "per sample".

**Comment by gbazad93** [comment]
> 
> 
> > So, as @charleygros suggested it can be optional for specific dataset and necessary for the test set.
> 
> Agree with this suggestion, let's do it :-)
> 
> Note that we have something similar already implemented at test time: the [evaluation pipeline on the 3D reconstructed images](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/evaluation.py#L114) --> which computes the evaluation metrics on the full 3D MRI image, instead of "per batch" or "per sample".

@charleygros now it is optional. Yes just noticed the evaluation pipeline is already considered this case. Thanks

File: ivadomed/losses.py [hunk 1]
```diff
@@ -51,16 +51,19 @@ class DiceLoss(nn.Module):
     Attributes:
         smooth (float): Value to avoid division by zero when images and predictions are empty.
```
Comment (hermancollin) [inline ivadomed/losses.py:52]:
```suggestion
        smooth (float): Value to avoid division by zero when images and predictions are empty.
        sample_wise (bool): If set, dice loss is averaged over sample instead of over batch
```

**Comment by charleygros** [comment]
> @charleygros now it is optional. Yes just noticed the evaluation pipeline is already considered this case. Thanks

Looks good to me!

Would be great to have someone else review before merging as it is quite a crucial part of the ivadomed creature, @uzaymacar @mariehbourget @andreanne-lemay ? :-)

**Comment by rezazad68** [comment]
@gbazad93 implementation looks good. I only concern about the extra computation burden. Since the sample-wise loss only applies in rare cases, the current implementation adds `two unsqueeze and one average operation`, which is unnecessary in the batch cases (consider that the loss needs to be calculated for each batch). Simple conditions and grouping into two separate parts might be computationally efficient.

**Comment by mariehbourget** [comment]
Thank you @gbazad93 for this contribution.
I tested it and everything looks good.
As @rezazad68 suggested, it may be worth it to test the computation burden.
Regarding the new option, @charleygros is sample_wise an optional parameter that we would want to add to the config file and the entire pipeline (with default behavior sample_wise=False)?

**Comment by charleygros** [comment]
>  Regarding the new option, @charleygros is `sample_wise` an optional parameter that we would want to add to the config file and the entire pipeline (with default behavior `sample_wise=False`)?

Happy either ways.
I think we don't need to make it appear in the config files etc. since that's quite an advanced parameter, and will appear in the online doc anyway.
But please have the conversation during meeting if unsure :-)

**Comment by mariehbourget** [comment]
> I think we don't need to make it appear in the config files etc. since that's quite an advanced parameter, and will appear in the online doc anyway.

My only concern about not having it in the config file is that the parameter becomes "invisible" and could affect reproducibility of the experiments.

**Comment by gbazad93** [comment]
Hi @charleygros and @uzaymacar, I hope you're doing well! Just a friendly reminder about this pull request. Please let me know if there’s anything else I can address to move it forward. Thanks!

**Comment by jcohenadad** [comment]
@gbazad93 ivadomed is not maintained anymore, sorry for the trouble.
